src/k2/httpproxy/HTTPProxy.cpp

```diff
@@ -362,6 +362,125 @@ seastar::future<std::tuple<Status, CollectionCreateResponse>> HTTPProxy::_handle
+    }
+}
+
+// Convert [{type: "FieldType", value: "value"}..] to string that can be used in
+// collection range end
+seastar::future<nlohmann::json> HTTPProxy::_handleGetKeyString(nlohmann::json&& request) {
+    String output;
+    if (!request.contains("fields"))
+        return JsonResponse(Status{400, "Invalid json"});
+
+    for (auto& record : request["fields"]) {
+        SchemaField field;
+        record.at("type").get_to(field.type);
+        String out;
+        K2_DTO_CAST_APPLY_FIELD_VALUE(getEscapedString, field, record["value"], out);
+        output += out;
+    }
+    return JsonResponse(Status{200, "success"}, nlohmann::json{{"result", output}});
+}
+
+
+seastar::future<nlohmann::json> HTTPProxy::_handleCreateQuery(nlohmann::json&& jsonReq) {
+    std::string collectionName;
+    std::string schemaName;
+
+    try {
+        jsonReq.at("collectionName").get_to(collectionName);
+        jsonReq.at("schemaName").get_to(schemaName);
+    } catch(...) {
+        return JsonResponse(Status{400, "Bad json for query request"});
+    }
+
+    return _client.createQuery(collectionName, schemaName)
+    .then([this, req=std::move(jsonReq)] (auto&& result) mutable {
+        if(!result.status.is2xxOK()) {
+            return JsonResponse(std::move(result.status));
+        }
+        K2LOG_D(log::httpproxy, "begin query {}", result);
+        if (req.contains("startScanRecord")) {
+            serializeRecordFromJSON(result.query.startScanRecord, std::move(req.at("startScanRecord")));
+        }
+        if (req.contains("endScanRecord")) {
+            serializeRecordFromJSON(result.query.endScanRecord, std::move(req.at("endScanRecord")));
+        }
+        if (req.contains("limit")) {
+            result.query.setLimit(req["limit"]);
+        }
+        if (req.contains("reverse")) {
+            result.query.setReverseDirection(req["reverse"]);
+        }
+        _queries[_queryID++] = std::move(result.query);
+        nlohmann::json resp{{"queryID", _queryID - 1}};
+        return JsonResponse(std::move(result.status), std::move(resp));
+    });
+}
+
+seastar::future<nlohmann::json> HTTPProxy::_handleQuery(nlohmann::json&& jsonReq) {
+    uint64_t txnID;
+    uint64_t queryID;
+
+    try {
+        jsonReq.at("txnID").get_to(txnID);
+        jsonReq.at("queryID").get_to(queryID); 
+    } catch(...) {
+        return JsonResponse(Status{400, "Bad json for query request"});
+    }
+    if (_txns.find(txnID) == _txns.end()) {
+        return JsonResponse(Status{400, "Could not find txnID for query request"});
+    }
+    if (_queries.find(queryID) == _queries.end()) {
+        return JsonResponse(Status{400, "Could not find queryID for query request"});
+    }
+
+    return _txns[txnID].query(_queries[queryID])
+    .then([this, queryID](QueryResult&& result) {
+        if(!result.status.is2xxOK()) {
+            return JsonResponse(std::move(result.status));
+        }
+        std::vector<nlohmann::json> records; 
```

Comment:
Can use reserve on this vector since we know the size before hand

Reply:
Yes. Done.